server/manager/src/main/java/org/apache/accumulo/manager/tableOps/Utils.java

```diff
@@ -65,15 +70,43 @@ public class Utils {
   public static void checkTableNameDoesNotExist(ServerContext context, String tableName,
       TableId tableId, TableOperation operation) throws AcceptableThriftTableOperationException {
 
+    final Map<NamespaceId,String> namespaces = new HashMap<>();
+    final boolean namespaceInTableName = tableName.contains(".");
     try {
       for (String tid : context.getZooReader()
           .getChildren(context.getZooKeeperRoot() + Constants.ZTABLES)) {
-        String zTablePath = context.getZooKeeperRoot() + Constants.ZTABLES + "/" + tid;
+
+        final String zTablePath = context.getZooKeeperRoot() + Constants.ZTABLES + "/" + tid;
         try {
-          byte[] tname = context.getZooReader().getData(zTablePath + Constants.ZTABLE_NAME);
+          final byte[] tname = context.getZooReader().getData(zTablePath + Constants.ZTABLE_NAME);
           Preconditions.checkState(tname != null, "Malformed table entry in ZooKeeper at %s",
               zTablePath);
-          if (tableName.equals(new String(tname, UTF_8)) && !tableId.equals(TableId.of(tid))) {
+
+          String namespaceName = Namespace.DEFAULT.name();
+          if (namespaceInTableName) {
+            final byte[] nId =
+                context.getZooReader().getData(zTablePath + Constants.ZTABLE_NAMESPACE);
+            if (nId != null) {
+              final NamespaceId namespaceId = NamespaceId.of(new String(nId, UTF_8));
+              if (!namespaceId.equals(Namespace.DEFAULT.id())) {
+                namespaceName = namespaces.get(namespaceId);
+                if (namespaceName == null) {
+                  try {
+                    namespaceName = Namespaces.getNamespaceName(context, namespaceId);
+                    namespaces.put(namespaceId, namespaceName);
+                  } catch (NamespaceNotFoundException e) {
+                    throw new AcceptableThriftTableOperationException(null, tableName,
+                        TableOperation.CREATE, TableOperationExceptionType.OTHER,
+                        "Table (" + tableId.canonical() + ") contains reference to namespace ("
+                            + namespaceId + ") that doesn't exist");
+                  }
+                }
+              }
+            }
+          }
+
+          if (tableName.equals(TableNameUtil.qualified(new String(tname, UTF_8), namespaceName))
```

Comment:
Experimented w/ consolidating code and it does not seem worthwhile because of difference in checked exceptions and slight diffs in behavior when reading from zoocache vs zookeeper (return null vs throw no node exception).

Reply:
The other places in the code that I found that running a similar algorithm are TableMap and ZooPropUtils.

Reply:
This is the experimental code that I abandoned.  The following was trying to make the code in TableMap more generic so it could be used in other places.    Was working in TableMap class, following does not compile though.

```java
class TableMap {
 @FunctionalInterface
  public static interface ZooData {
    byte[] getData(String path) throws KeeperException, InterruptedException;
  }

  @FunctionalInterface
  public static interface TableIdConsumer {
    public void accept(TableId tableId, String name) throws Exception;
  }

  public static void readTableNames(String zkRoot, List<String> tableIds, ZooData zookeeper, TableIdConsumer tableConsumer) throws Exception{

    // use StringBuilder to construct zPath string efficiently across many tables
    StringBuilder zPathBuilder = new StringBuilder();
    zPathBuilder.append(zkRoot).append(Constants.ZTABLES).append("/");
    int prefixLength = zPathBuilder.length();

    Map<NamespaceId,String> namespaceIdToNameMap = new HashMap<>();

    for (String tableIdStr : tableIds) {
      // reset StringBuilder to prefix length before appending ID and suffix
      zPathBuilder.setLength(prefixLength);
      zPathBuilder.append(tableIdStr).append(Constants.ZTABLE_NAME);
      byte[] tableName = zookeeper.getData(zPathBuilder.toString());
      zPathBuilder.setLength(prefixLength);
      zPathBuilder.append(tableIdStr).append(Constants.ZTABLE_NAMESPACE);
      byte[] nId = zookeeper.getData(zPathBuilder.toString());

      String namespaceName = Namespace.DEFAULT.name();
      // create fully qualified table name
      if (nId == null) {
        namespaceName = null;
      } else {
        NamespaceId namespaceId = NamespaceId.of(new String(nId, UTF_8));
        if (!namespaceId.equals(Namespace.DEFAULT.id())) {
          try {
            namespaceName = namespaceIdToNameMap.get(namespaceId);
            if (namespaceName == null) {
              namespaceName = Namespaces.getNamespaceName(context, namespaceId);
              namespaceIdToNameMap.put(namespaceId, namespaceName);
            }
          } catch (NamespaceNotFoundException e) {
            log.error("Table (" + tableIdStr + ") contains reference to namespace (" + namespaceId
                    + ") that doesn't exist", e);
            continue;
          }
        }
      }
      if (tableName != null && namespaceName != null) {
        String tableNameStr = TableNameUtil.qualified(new String(tableName, UTF_8), namespaceName);
        TableId tableId = TableId.of(tableIdStr);
        tableConsumer.accept(tableId, tableNameStr);
      }
    }
  }
}
```